src/Modules/CalcDefence.lua

```diff
@@ -1661,8 +1661,156 @@ function calcs.defence(env, actor)
+				local tempDamage = m_min(damageRemainder * (1 - esBypass / 100), energyShield)
+				energyShield = energyShield - tempDamage
+				damageRemainder = damageRemainder - tempDamage
+			end
+			if (output.sharedMindOverMatter + output[damageType.."MindOverMatter"]) > 0 then
+				local MoMDamage = damageRemainder * m_min(output.sharedMindOverMatter + output[damageType.."MindOverMatter"], 100) / 100
+				if modDB:Flag(nil, "EnergyShieldProtectsMana") and energyShield > 0 and esBypass < 100 then
+					local tempDamage = m_min(MoMDamage * (1 - esBypass / 100), energyShield)
+					energyShield = energyShield - tempDamage
+					MoMDamage = MoMDamage - tempDamage
+					local tempDamage2 = m_min(MoMDamage, mana)
+					mana = mana - tempDamage2
+					damageRemainder = damageRemainder - tempDamage - tempDamage2
+				elseif mana > 0 then
+					local tempDamage = m_min(MoMDamage, mana)
+					mana = mana - tempDamage
+					damageRemainder = damageRemainder - tempDamage
+				end
+			end
+			if output.preventedLifeLossTotal > 0 then
+				if output.preventedLifeLossBelowHalf ~= 0 then
+					local lifeOverLowLife = m_max(life - output.Life * data.misc.LowPoolThreshold, 0)
+					local preventPercent = output.preventedLifeLoss / 100
+					local poolAboveLow = lifeOverLowLife / (1 - preventPercent)
+					local damageToSplit = m_min(damageRemainder, poolAboveLow)
+					local lostLife = damageToSplit * (1 - preventPercent)
+					local preventedLoss = damageToSplit * preventPercent
+					damageRemainder = damageRemainder - damageToSplit
+					LifeLossLostOverTime = LifeLossLostOverTime + preventedLoss
+					life = life - lostLife
+					
+					if life <= output.Life * data.misc.LowPoolThreshold then
+						local unspecificallyLowLifePreventedDamage = damageRemainder * preventPercent
+						LifeLossLostOverTime = LifeLossLostOverTime + unspecificallyLowLifePreventedDamage
+						damageRemainder = damageRemainder - unspecificallyLowLifePreventedDamage
+						local specificallyLowLifePreventedDamage = damageRemainder * modDB:Sum("BASE", nil, "LifeLossBelowHalfPrevented") / 100
+						LifeBelowHalfLossLostOverTime = LifeBelowHalfLossLostOverTime + specificallyLowLifePreventedDamage
+						damageRemainder = damageRemainder - specificallyLowLifePreventedDamage
+					end
+				else
+					local tempDamage = damageRemainder * output.preventedLifeLoss / 100
+					LifeLossLostOverTime = LifeLossLostOverTime + tempDamage
+					damageRemainder = damageRemainder - tempDamage
+				end
+			end
+			life = life - damageRemainder
+		end
+		
+		return {
+			FrostShieldLife = frostShield,
+			SoulLink = soulLink,
+			Aegis = aegis,
+			Guard = guard,
+			PoolsLost = PoolsLost,
+			Ward = restoreWard,
+			EnergyShield = energyShield,
+			Mana = mana,
+			Life = life,
+			LifeLossLostOverTime = LifeLossLostOverTime,
+			LifeBelowHalfLossLostOverTime = LifeBelowHalfLossLostOverTime
+		}
+	end
+	output.reducePoolsByDamage = reducePoolsByDamage
```

Comment:
I dont know how much this is done elsewhere, but functions should probably not be inserted directly into output, this seems to just be used for tests atm?

imo it should either be referenced elsewhere (like in "calcs.reducePoolsByDamage"), or at the very least under an "output.functions.reducePoolsByDamage" so that all the functions in output are grouped together

otherwise the the rest of this PR seems fine, it might be a decent idea to move both the helper functions outside of the main function (before line 52) but it should be fine to merge as is